### DISCUSSION TITLE
RLC-820A via NVR

### DISCUSSION THREAD

**Discussion by vsurao** [thirtythreeforty/neolink#291] [General]
Hello, want to understand something. I have an existing system with 4 D800 cameras connected to an RLN8-410. Neolink works great with these. 
I recently got an additional RLC-820A. I understand this is an RTSP camera, however I am still using the NVR for redundancy so wanted to connect to the camera off the NVR similar to the D800s.

Now first of all, I did get connect successfully to the camera via the NVR using channel id just like I did for the D800s. However I noticed a few things.
 1. Whenever the machine hosting Neolink was rebooted, Neolink would give me deserialisation, I/O and connection refused errors for about 10 minutes or so. Eventually it would settle and all 5 cameras (4 D800 + 1 RLC820A) would connect. 
2. What I also noticed was that if I knocked off any 1 camera from my list of 5 cameras in the 'toml' file, Neolink would connect to all 4 instantly. If I included the 5th (either a D800 or the RLC 820A) the errors would return for 10 minute or whatever. 

So 2 questions:
1. Is my understanding correct that via the Reolink NVR, it doesnt matter which camera you connect (RTSP capable or not) the dual stream required for third party software like BlueIris will be the non-RTSP (Baichuan) protocol stream?.
2. Why is there this delay in being able to get Neolink connected to the NVR when I have the 5 cameras?. 

**error message: (get same error block for all 5 cameras)**
[2022-07-21T14:07:17Z INFO  neolink::rtsp] back_yard: Connecting to camera at Address: 192.168.0.206:9000
[2022-07-21T14:07:17Z INFO  neolink::rtsp] back_yard: Logging in
[2022-07-21T14:07:17Z ERROR neolink_core::bc_protocol::connection::bcconn] Deserialization error: Deserialization error
[2022-07-21T14:07:17Z ERROR neolink_core::bc_protocol::connection::bcconn] caused by: I/O error
[2022-07-21T14:07:17Z ERROR neolink_core::bc_protocol::connection::bcconn] caused by: An established connection was aborted by the software in your host machine. (os error 10053)
[2022-07-21T14:07:17Z ERROR neolink::rtsp] Error streaming from camera back_yard, will retry in 8s: Failed to login to back_yard

    Caused by:
        0: Connection error
        1: Serialization error
        2: IoError(Os { code: 10053, kind: ConnectionAborted, message: "An established connection was aborted by the software in your host machine." })

**Comment by QuantumEntangledAndy** [comment]
I'll try and answer your questions but I may not be understanding everything.

1. I believe this is true, internally reolink uses the Baichuan protocol to forward all of its command and control messages as well as its streams (slightly different format and magic header though)
2. Not sure, this could be to do with how neolink tries to estabish a constant simultaneous connection and it is saturating the network. If you have both the HD and SD streams on this will mean 10 simultaneous streams, which is probably a lot of data. You could perhaps test this by enabling just the subSteams in the config files.

**Reply by vsurao** [reply]
thank you for the reply, tried what you suggested and something interesting happened.
1. The substream only setup connected instantly
2. Once the substream only setup connected, I could terminate it immediately and establish a dual streaming connection

In my very layman language, it appears the issue was authenticating across 5 cameras when dual streams were active for all (too many requests going in too fast?).
Once the host machine was authenticated using the sub-stream , the NVR seemed to accept dual stream connections from Neolink just fine.

**Reply by QuantumEntangledAndy** [reply]
It is possible that the NVR and the network gets saturated. I have some ideas to relieve this load, but I just don't have the time really to update neolink these days :(

**Reply by vsurao** [reply]
Totally understand, I am going to modify the task scheduler to launch 2 Neolink instance sin sequence, hopefully that should be a simple fix. Dirty, but may work. Will try and report back for all its worth

**Reply by twistedddx** [reply]
Have you tried the single_login_stream branch?
4x hd streams on the original connection method was all nvr's would handle before issues start.
https://github.com/QuantumEntangledAndy/neolink/actions/runs/2558840040
